verification/abi.go

```diff
@@ -126,3 +154,370 @@ type GetTaggedTCIResp[Digest DigestAlgorithm] struct {
 	CumulativeTCI Digest
 	CurrentTCI    Digest
 }
+
+// DPEABI is a connection to a DPE instance, parameterized by hash algorithm and ECC curve.
+type DPEABI[CurveParameter Curve, Digest DigestAlgorithm] struct {
+	transport    Transport
+	Profile      Profile
+	MajorVersion uint16
+	MinorVersion uint16
+	VendorId     uint32
+	VendorSku    uint32
+	MaxTciNodes  uint32
+	Flags        uint32
+}
+
+// DPEABI256 is a client that implements DPE_PROFILE_IROT_P256_SHA256
+type DPEABI256 = DPEABI[NISTP256Parameter, SHA256Digest]
+
+// DPEABI384 is a client that implements DPE_PROFILE_IROT_P384_SHA384
+type DPEABI384 = DPEABI[NISTP384Parameter, SHA384Digest]
+
+// dpeProfileImplementsTypeConstraints checks that the requested DPEABI type constraints are compatible with the DPE profile.
+func dpeProfileImplementsTypeConstraints[C Curve, D DigestAlgorithm](profile Profile) error {
+	// Test that the expected value types produced by each DPE profile can be assigned to variables of type C and D
+	var c C
+	var d D
+	switch profile {
+	case ProfileP256SHA256:
+		// We must cast c and d to any in order to perform type assertions on them.
+		// https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md#why-not-permit-type-assertions-on-values-whose-type-is-a-type-parameter
+		if _, ok := any(c).(NISTP256Parameter); !ok {
+			return fmt.Errorf("an incorrect ECC parameter type was passed to a DPE implementing DPE_PROFILE_IROT_P256_SHA256")
+		}
+		if _, ok := any(d).(SHA256Digest); !ok {
+			return fmt.Errorf("an incorrect digest type was passed to a DPE implementing DPE_PROFILE_IROT_P256_SHA256")
+		}
+		return nil
+	case ProfileP384SHA384:
+		if _, ok := any(c).(NISTP384Parameter); !ok {
+			return fmt.Errorf("an incorrect ECC parameter type was passed to a DPE implementing DPE_PROFILE_IROT_P384_SHA384")
+		}
+		if _, ok := any(d).(SHA384Digest); !ok {
+			return fmt.Errorf("an incorrect digest type was passed to a DPE implementing DPE_PROFILE_IROT_P384_SHA384")
+		}
+		return nil
+	}
+	return fmt.Errorf("unsupported DPE profile: %v", profile)
+}
+
+// NewDPEABI initializes a new DPE client.
+func NewDPEABI[C Curve, D DigestAlgorithm](t Transport) (*DPEABI[C, D], error) {
```

Comment:
optional: something to think about (but not for too long, this is really an optional comment): do you expect callers to actually call `NewDPEABI` or `NewDPEABI256` or `NewDPEABI384`, or do you expect them to call `NewClient` and then perform a type assertion that they got e.g., a `DPEABI256`? (Maybe you expect some will do the first thing, some will do the second thing).

Depending on how opinionated you feel about what callers should do, you can hide some of these symbols:

- the actual parameterized DPEABI type, renaming to dpeABI (leaving the aliases `DPEABI256` and `DPEABI384` exported for callers who want to write code for *their* expected profile)
- the `NewDPEABI*` factories (requiring the caller to ask the client library which DPE ABI it's got)

In general if this library gets some (non-external, non-test) callers, it'd be easier to make private things public rather than the other way around, so I'd suggest making any of this stuff private that's not needed to write the tests.

Reply:
I could see a caller which really does want a 256 client and wants to fail if that's not what the RoT implements. So I'll leave the NewDPEABI* factories. But I changed the type-parameterized DPEABI to be private.